### PR TITLE
Support transaction for the rest of APIs in task services

### PR DESCRIPTION
**Pull request by Junoth**
Tested in Imsonia
Screen Shot 2023-03-02 at 7 35 18 PM
Screen Shot 2023-03-02 at 7 35 01 PM
Screen Shot 2023-03-02 at 7 30 29 PM
Screen Shot 2023-03-02 at 7 28 59 PM
Screen Shot 2023-03-02 at 7 26 49 PM
Screen Shot 2023-03-02 at 7 26 28 PM
Screen Shot 2023-03-02 at 7 24 55 PM
Screen Shot 2023-03-02 at 7 24 12 PM

### CODE REVIEW

File: core/realtime/mutation.go [hunk 1]
```diff
@@ -22,7 +22,7 @@ type Mutation interface {
 	ExecuteV2(ct context.Context, tx *sql.Tx) *errs.Error
 	Undo() *errs.Error
 	CleanUp(ct context.Context) *errs.Error
-	// Deprecated: The old method should not be used anymore. Use another PrepareClientNotifiers method instead
+	// Deprecated: The old method should not be used anymore. Use another PrepareClientNotifiers and GetClientNotifiersV2 method instead
```
Comment (naturetime10) [inline core/realtime/mutation.go]:
```suggestion
	// Deprecated: use PrepareClientNotifiers and GetClientNotifiersV2 method instead
```

File: core/service/task.go [hunk 2]
```diff
@@ -67,24 +67,53 @@ type Task struct {
...
+	var tasks []entity.Task
+	txCtx := TransactionsContext{
+		dataCollector: t.dataCollector,
+		db:            t.db,
+		stateSyncer:   t.stateSyncer,
+		ct:            ct,
+	}
+	err := txCtx.withTransactions(true, func(sqlTx *sql.Tx, rtTx *realtime.Transaction) *errs.Error {
+		var internalErr *errs.Error
+		tasks, internalErr = t.taskDaoV2.FindAllTasks(ct, sqlTx)
```
Comment (naturetime10) [inline core/service/task.go:113]: Should we log error for stack trace? Are we waiting for the `error stacktrace` PR instead?

File: core/service/task.go [hunk 3]
```diff
@@ -98,7 +127,19 @@ func (t Task) FindTasks(ct context.Context, filter *TaskFilter) ([]entity.Task,
...
+	var tasks []entity.Task
+	txCtx := TransactionsContext{
+		dataCollector: t.dataCollector,
+		db:            t.db,
+		stateSyncer:   t.stateSyncer,
+		ct:            ct,
+	}
+	err := txCtx.withTransactions(true, func(sqlTx *sql.Tx, rtTx *realtime.Transaction) *errs.Error {
+		var internalErr *errs.Error
+		tasks, internalErr = t.taskDaoV2.FindTasksByTeamID(ct, sqlTx, teamID)
```
Comment (naturetime10) [inline core/service/task.go:139]: Should we log error for stack trace? Are we waiting for the `error stacktrace` PR instead?

File: core/service/task.go [hunk 4]
```diff
@@ -116,13 +157,24 @@ func (t Task) FindTasksInSprint(
...
+		ct:            ct,
 	}
+	err := txCtx.withTransactions(true, func(sqlTx *sql.Tx, rtTx *realtime.Transaction) *errs.Error {
+		taskIDs, internalErr := t.sprintTaskRelationDaoV2.FindTaskIDsBySprintID(ct, sqlTx, sprintID)
+		if internalErr != nil {
+			t.dataCollector.Logger.ErrorWithContext(ct, internalErr)
+			return internalErr
+		}
+
+		tasks, internalErr = t.taskDaoV2.FindTasksByIDs(ct, sqlTx, taskIDs)
```
Comment (naturetime10) [inline core/service/task.go:174]: Should we log error for stack trace? Are we waiting for the `error stacktrace` PR instead?

File: core/service/task.go [hunk 5]
```diff
@@ -136,19 +188,30 @@ func (t Task) FindTasksInSprint(
...
+		ct:            ct,
 	}
+	err := txCtx.withTransactions(true, func(sqlTx *sql.Tx, rtTx *realtime.Transaction) *errs.Error {
+		taskIDs, internalErr := t.taskAwaitForRelationDaoV2.FindAwaitForTaskIDs(ct, sqlTx, awaitingTaskID)
+		if internalErr != nil {
+			t.dataCollector.Logger.ErrorWithContext(ct, internalErr)
+			return internalErr
+		}
+
+		tasks, internalErr = t.taskDaoV2.FindTasksByIDs(ct, sqlTx, taskIDs)
```
Comment (naturetime10) [inline core/service/task.go:205]: Should we log error for stack trace? Are we waiting for the `error stacktrace` PR instead?

File: core/service/task.go [hunk 6]
```diff
@@ -171,57 +233,79 @@ func (t Task) createTask(ct context.Context, teamID uint64, taskInput createTask
...
+			t.taskDaoV2,
+			task,
+		)
+
+		internalErr = createTaskMutation.ExecuteV2(ct, sqlTx)
+		if internalErr != nil {
+			t.dataCollector.Logger.ErrorWithContext(ct, internalErr)
+			return internalErr
+		}
+		rtTx.AppendMutation(createTaskMutation)
```
Comment (naturetime10) [inline core/service/task.go]:
```suggestion
		}
		
		rtTx.AppendMutation(createTaskMutation)
```

File: core/service/task.go [hunk 7]
```diff
@@ -171,57 +233,79 @@ func (t Task) createTask(ct context.Context, teamID uint64, taskInput createTask
...
+
```
Comment (naturetime10) [inline core/service/task.go:288]:
```suggestion
```

File: core/service/task.go [hunk 8]
```diff
@@ -608,53 +694,60 @@ func (t Task) moveTaskToUpcoming(ct context.Context, tx *realtime.Transaction, t
 		t.taskDaoV2,
 		task,
 	)
-	err = tx.ApplyMutation(ct, updateTaskMutation)
-	if err != nil {
-		t.dataCollector.Logger.ErrorWithContext(ct, err)
-		return entity.Task{}, err
+	rtTx.AppendMutation(updateTaskMutation)
```
Comment (naturetime10) [inline core/service/task.go:697]: Can we add an empty line between line `686` and `687`? Thanks!

File: core/service/task.go [hunk 9]
```diff
@@ -608,53 +694,60 @@ func (t Task) moveTaskToUpcoming(ct context.Context, tx *realtime.Transaction, t
...
+		if autoPauseTask {
+			switch task.Status {
+			case entity.TaskStatusInProgress, entity.TaskStatusPaused:
+				task.Status = entity.TaskStatusPaused
+			}
+		} else {
+			task.Status = entity.TaskStatusTodo
+		}
+		now := time.Now()
+		task.UpdatedAt = &now
```
Comment (naturetime10) [inline core/service/task.go]:
```suggestion
		}
		
		now := time.Now()
		task.UpdatedAt = &now
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]
Can we rename this PR's title to `Support transaction for the rest of task service APIs`?

File: core/service/task.go (hunk 4 above)
Comment (Junoth) [reply core/service/task.go:174]: Yeah. I'll do that after error stack trace PR

File: core/service/task.go (hunk 8 above)
Comment (Junoth) [reply core/service/task.go:697]: sure

**Review by naturetime10 — APPROVED** [review]

**Pull request merged**